### PR TITLE
Move to a structured status for dynamic kubelet config

### PR DESCRIPTION
**Pull request by mtaufen**
This PR updates dynamic Kubelet config to use a structured status, rather than a node condition. This makes the status machine-readable, and thus more useful for config orchestration. 

Fixes: #56896

```release-note
The status of dynamic Kubelet config is now reported via Node.Status.Config, rather than the KubeletConfigOk node condition.
```

### CODE REVIEW

File: pkg/kubelet/kubeletconfig/checkpoint/download.go [hunk 1]
```diff
@@ -46,39 +46,35 @@ type Payload interface {
...
-	// at this point we know we're using the ConfigMapRef subfield
-	ref := source.ConfigMapRef
+	// NOTE: Even though the API server validates the config, we check whether all *known* fields are
+	// nil here, so that if a new API server allows a new config source type, old clients can send
+	// an error message rather than crashing due to a nil pointer dereference.
 
-	// name, namespace, and UID must all be non-empty for ConfigMapRef
-	if ref.Name == "" || ref.Namespace == "" || string(ref.UID) == "" {
-		return nil, status.FailSyncReasonPartialObjectReference, fmt.Errorf("%s, ObjectReference was: %#v", status.FailSyncReasonPartialObjectReference, ref)
+	// exactly one reference subfield of the config source must be non-nil, toady ConfigMap is the only reference subfield
```
[the remark this answers is not in the document]
Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/checkpoint/download.go]: updated, thanks

File: pkg/kubelet/kubeletconfig/controller.go [hunk 2]
```diff
@@ -79,59 +74,74 @@ func NewController(defaultConfig *kubeletconfig.KubeletConfiguration, dynamicCon
...
 	}
 
-	// set the status to indicate that we had to roll back to the lkg for the reason reported when we tried to load the assigned config
-	if lkgSource != nil {
-		cc.configOk.Set(fmt.Sprintf(status.LkgRemoteMessageFmt, lkgSource.APIPath()), reason, apiv1.ConditionFalse)
-	} else {
-		cc.configOk.Set(status.LkgLocalMessage, reason, apiv1.ConditionFalse)
+	// attempt to load the last-known-good config
+	lastKnownGoodConfig, _, err := cc.loadConfig(lastKnownGoodSource)
+	if err == nil {
```
[the remark this answers is not in the document]
Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/controller.go]: Updated, thanks.

File: pkg/apis/core/types.go [hunk 3]
```diff
@@ -3348,6 +3348,23 @@ type NodeSystemInfo struct {
 	Architecture string
 }
 
+// NodeConfigStatus describes the status of the config assigned by Node.Spec.ConfigSource
+type NodeConfigStatus struct {
+	// Assigned is the node's ack of the config source assigned via Node.Spec.ConfigSource,
+	// and is a direct reflection of that source.
+	// +optional
+	Assigned *NodeConfigSource
```
Comment (liggitt) [inline pkg/apis/core/types.go]: can you clarify the circumstances when Assigned != Active, and when Active != LastKnownGood?

File: pkg/registry/core/node/strategy.go [hunk 4]
```diff
@@ -127,8 +127,14 @@ type nodeStatusStrategy struct {
 var StatusStrategy = nodeStatusStrategy{Strategy}
 
 func (nodeStatusStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
-	_ = obj.(*api.Node)
+	node := obj.(*api.Node)
 	// Nodes allow *all* fields, including status, to be set on create.
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DynamicKubeletConfig) {
+		node.Status.Config.Active = nil
```
Comment (liggitt) [inline pkg/registry/core/node/strategy.go]: seems like the whole config field should disappear in that case

Comment (liggitt) [reply pkg/registry/core/node/strategy.go]: also need to strip this field on update

File: pkg/apis/core/types.go (hunk 3 above)
Comment (liggitt) [reply pkg/apis/core/types.go]: for example, it would be helpful to know things like this (these are just my imaginings, review for accuracy and for the implications of those states):

1. when the node has observed an updated config, has validated it, but has not yet made it active, Assigned will reflect the new config and Active will reflect the previous, still-running config. The node will not attempt to use the referenced config again until it is updated.
2. when the node has observed an updated config, cannot validate it, and will not make it active, Assigned will reflect the new config, Active will reflect the previous, still-running config, and Error will contain the validation error. The node will not attempt to use the referenced config again until it is updated.
3. when the node started with an updated config, encountered errors, and reverted to a LastKnownGood config, Assigned will reflect the new config, Active will contain the LastKnownGood config, and Error will contain a description of the error encountered. The node will not attempt to use the referenced config again until it is updated.
4. when the node has started with an updated config, but has not yet run long enough to verify the config is good, LastKnownGood will reference the previous known good config (if any exists).

File: pkg/apis/core/types.go [hunk 5]
```diff
@@ -3348,6 +3348,23 @@ type NodeSystemInfo struct {
...
+	// Active reports the config source the node is actively using
+	// +optional
+	Active *NodeConfigSource
+	// LastKnownGood is the config source that will be used if an invalid config is assigned in the future
+	// +optional
+	LastKnownGood *NodeConfigSource
+	// Error describes any problems with the configuration
+	// +optional
+	Error string
```
Comment (liggitt) [inline pkg/apis/core/types.go]: clarify whether the presence of Error means the Assigned config source will be ignored until changes are made to it

File: pkg/apis/core/types.go [hunk 6]
```diff
@@ -3380,6 +3397,9 @@ type NodeStatus struct {
 	// List of volumes that are attached to the node.
 	// +optional
 	VolumesAttached []AttachedVolume
+	// Status of the config assigned to the node
```
Comment (liggitt) [inline pkg/apis/core/types.go]: clarify this only applies if dynamic config feature is used and a ConfigSource is specified for the node?

File: pkg/apis/core/validation/validation.go [hunk 7]
```diff
@@ -4162,28 +4171,74 @@ func validateConfigMapNodeConfigSourceSpec(source *core.ConfigMapNodeConfigSourc
 	return append(allErrs, validateConfigMapNodeConfigSource(source, fldPath)...)
 }
 
+// validation specififc to Node.Status.Config
+func validateNodeConfigStatus(status *core.NodeConfigStatus, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if status.Assigned != nil {
+		// Assigned is a reflection of the spec, which acts as an ACK from the Kubelet regarding recognition,
+		// of the assignment. As such, it is validated against the same rules as the spec, rather than status.
```
Comment (liggitt) [inline pkg/apis/core/validation/validation.go]: I thought you wanted resource version and uid reflected in the status, to know whether the node had observed a particular version of the source

File: pkg/apis/core/validation/validation.go [hunk 8]
```diff
@@ -4162,28 +4171,74 @@ func validateConfigMapNodeConfigSourceSpec(source *core.ConfigMapNodeConfigSourc
...
+
+// validation specific to Node.Status.Config.(Active|Assigned|LastKnownGood).ConfigMap
+func validateConfigMapNodeConfigSourceStatus(source *core.ConfigMapNodeConfigSource, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if string(source.UID) == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("uid"), "uid must be set in status"))
+	}
+	// TODO(#63221): require ResourceVersion in status when we start respecting ConfigMap mutations (the Kubelet isn't tracking it internally until
+	// that PR, which makes it difficult to report for now).
```
Comment (liggitt) [inline pkg/apis/core/validation/validation.go]: even if we don't track mutations, we should still have the actual configmap in hand when we report status, right? why can't we report the resourceVersion we're using?

File: pkg/apis/core/validation/validation.go (hunk 7 above)
Comment (liggitt) [reply pkg/apis/core/validation/validation.go]: I suppose one possibility is we've seen the assigned reference and are getting an error fetching it, in which case we wouldn't have a uid or resourceVersion

File: pkg/apis/core/validation/validation.go (hunk 8 above)
Comment (mtaufen) [reply pkg/apis/core/validation/validation.go]: We set status when we bootstrap the controller, when an active config becomes the new last-known-good, and when we ack the assigned config. The last of those three doesn't need UID/ResourceVersion because it's a reflection of spec; but during the former two we only have metadata from SerializedNodeConfigSources. Until I made the changes in #63221 to respect ConfigMap updates, we weren't setting ResourceVersion in the SerializedNodeConfigSources (UID was the only identifier that mattered for resolving a checkpoint), so we didn't know it at bootstrap/lkg-update time. 

It wouldn't be too hard to port that fix back to this PR, though.

File: pkg/apis/core/types.go (hunk 6 above)
Comment (mtaufen) [reply pkg/apis/core/types.go]: will do

File: pkg/apis/core/validation/validation.go (hunk 7 above)
Comment (mtaufen) [reply pkg/apis/core/validation/validation.go]: > I thought you wanted resource version and uid reflected in the status, to know whether the node had observed a particular version of the source

For `Active` and `LastKnownGood`, yes.

`Assigned` is just an ack, so you know what the Kubelet thinks you told it to do. When we start responding to ConfigMap mutations, the meaning of `ConfigSource`, assigned in spec, will become "I'm assigning you the latest config at this namespace/name, invariant to UID and ResourceVersion," and I want the meaning of `Assigned` in status to be "I recognize that I should use the latest config at that namespace/name, invariant to UID and ResourceVersion."

> I suppose one possibility is we've seen the assigned reference and are getting an error fetching it, in which case we wouldn't have a uid or resourceVersion

Exactly, `Assigned` gets acknowledged before the Kubelet tries to download the config, and if an error is encountered while downloading, `Error` gets set subsequently.

File: pkg/registry/core/node/strategy.go (hunk 4 above)
Comment (mtaufen) [reply pkg/registry/core/node/strategy.go]: Good catch, I changed it to a pointer type midway through working on this and forgot to update this.

Comment (mtaufen) [reply pkg/registry/core/node/strategy.go]: updated PrepareForUpdate as well

File: pkg/apis/core/types.go (hunk 3 above)
Comment (mtaufen) [reply pkg/apis/core/types.go]: That's mostly correct. 

With (2), `Active` remains unchanged if the error happens prior to the Kubelet restarting to use the new config (e.g. some error prevented downloading or checkpointing), but will change to `LastKnownGood` if the Kubelet progresses to a restart and then encounters an error loading, parsing, or validating the unpacked payload (in the future, I'd like to optimize the flow by trying a load/parse/validate prior to restart, but this is how things work today).

File: pkg/apis/core/types.go (hunk 5 above)
Comment (mtaufen) [reply pkg/apis/core/types.go]: will do

File: pkg/apis/core/types.go (hunk 6 above)
Comment (mtaufen) [reply pkg/apis/core/types.go]: Technically it also applies when dynamic config is enabled, but you've set a nil configSource.

File: pkg/apis/core/types.go (hunk 3 above)
Comment (mtaufen) [reply pkg/apis/core/types.go]: Updated, let me know if you see other edges I should clarify.

File: pkg/apis/core/types.go [hunk 9]
```diff
@@ -3348,6 +3348,45 @@ type NodeSystemInfo struct {
...
 }
 
+// NodeConfigStatus describes the status of the config assigned by Node.Spec.ConfigSource.
+type NodeConfigStatus struct {
+	// Assigned reports the node's understanding of intended config. When the node detects an
+	// update to Node.Spec.ConfigSource, it updates Node.Status.Config.Assigned before
+	// taking further action. The node tries to make the Assigned config the Active config
+	// by downloading, checkpointing, validating, and attempting to use the referenced source.
+	// +optional
+	Assigned *NodeConfigSource
```
Comment (dashpole) [inline pkg/apis/core/types.go]: From our discussion offline: I think having transition times, and heartbeat times on each of these, showing when they were last updated, and when it last changed would be useful.  It would allow distinguishing between roll-backs to LKG and when the node config hasn't yet taken effect because if assigned has transitioned more recently than active, it hasn't taken effect, and if the opposite, it was rolled back.  Even though we have well defined errors, this seems a better way to transmit this information.  It can also help with correlating config changes with other, non-fatal behavior changes (e.g. lots of evictions start happening after enabling local storage).  Or, for example if the kubelet wasn't updating LKG after 10 minutes, or if the kubelet was in an infinite loop during loading of config, and heartbeat hadn't been updated in a couple minutes, these would be obvious.  I agree that having well defined objects in the status is helpful for this, but I think we should model this as a "typed condition", and keep the heartbeat and transition times for each of these Sources.

File: cmd/kubelet/app/server.go [hunk 10]
```diff
@@ -198,25 +198,43 @@ HTTP server: The kubelet can also listen for HTTP and respond to a simple API
...
-				// This is necessary to preserve backwards-compatibility across binary upgrades.
-				// See issue #56171 for more details.
-				if err := kubeletConfigFlagPrecedence(kubeletConfig, args); err != nil {
-					glog.Fatal(err)
-				}
-				// update feature gates based on new config
-				if err := utilfeature.DefaultFeatureGate.SetFromMap(kubeletConfig.FeatureGates); err != nil {
-					glog.Fatal(err)
+				// If we should just use our existing, local config, the controller will return a nil config
+				if dynamicKubeletConfig != nil {
```
Comment (dashpole) [inline cmd/kubelet/app/server.go]: nit: i'm generally not a fan of giving nil an implicit meaning.  I would rather have an explicit return value indicating if we should use local or remote.

Comment (mtaufen) [reply cmd/kubelet/app/server.go]: I'm not sure I agree in this case.
The controller bootstrap returns the dynamic config if it exists.
If there's no dynamic config to use it returns nil (`nil` => `Nothing` is a common idiom in Go). 
In the case that there's no dynamic config to use, we just move on and use the local.

File: pkg/apis/core/types.go (hunk 9 above)
Comment (mtaufen) [reply pkg/apis/core/types.go]: > distinguishing between roll-backs to LKG and when the node config hasn't yet taken effect 

To clarify, this is the scenario where we're on LKG, then ConfigSource is updated by a user, then Assigned is updated by the Kubelet, but we haven't tried the new config yet, so the status looks inconsistent (and it's unclear if Error refers to the new Assigned or the previous Assigned)?

And the transition times would clarify this by allowing you to compare the transition time for Assigned with the transition time for Error?

> It can also help with correlating config changes with other, non-fatal behavior changes

I wonder if some of these debugging scenarios aren't better covered by monitoring events or timeseries derived from metrics. We could send events at every status transition, rather than just when the Kubelet restarts to try a new config.

I want to be careful with heartbeats, as they do impact scalability (every update, including heartbeats, requires a full rewrite of the Node object in etcd). But I think transition times could provide some value.

> if the kubelet was in an infinite loop during loading of config

I think you'd already get a `NodeNotReady` in this case.

Comment (mtaufen) [inline pkg/apis/core/types.go]: TODO: Think through this case more:
1. API server is upgraded to version w/ new source subtype, Kubelets not
2. User sets new source type
3. Kubelet sets `Assigned` in runtime status manager, intending to ack, but had unmarshaled a config with all nil subfields, as far as it could see. So the status sync will fail API server validation.
4. Kubelet sees `AllNilSubfieldsError` when it tries to produce a config source a little after updating runtime status manager, and updates the status manager with this error. But since `Assigned` was already set to an invalid value in the status manager, all status sync attempts will fail API server validation until this is corrected.

Comment (mtaufen) [reply pkg/apis/core/types.go]: I spent the afternoon drawing up a state machine diagram, to help clarify what we should think about for status reporting: https://docs.google.com/drawings/d/1Xk_uiDFY0Y3pN6gualoy9wDPnja9BAuT-i5JYuAZ6wE/edit?usp=sharing

Comment (mtaufen) [reply pkg/apis/core/types.go]: I'm thinking that rather than having Assigned in the status be an early acknowledgement of ConfigSource, it might be clearer to make it a direct reflection of the fully explicit on-disk record of assigned config (like LKG is), and then ensure the error messages clearly differentiate between errors related to the Spec.ConfigSource vs errors related to the already-downloaded config payload.

In general, it's probably clearer if the status simply maps to the state machine.

Comment (mtaufen) [reply pkg/apis/core/types.go]: And if we do report timestamps, the simplest option might just be to report the modification times of the on-disk records for Assigned and LastKnownGood. (Active is a little trickier, since this is determined at Kubelet startup and only applies to the runtime; though the NodeReady heartbeat might be a decent proxy for whether Active is up to date or not).

File: cmd/kubelet/app/server.go (hunk 10 above)
Comment (dashpole) [reply cmd/kubelet/app/server.go]: sgtm

File: pkg/apis/core/types.go (hunk 9 above)
Comment (mtaufen) [reply pkg/apis/core/types.go]: @dashpole and I had an offline meeting and decided to leave timestamps out for now, and potentially add them later if a controller can justify that it needs them to reason about the state of the world.

Comment (mtaufen) [reply pkg/apis/core/types.go]: Now that we report the checkpointed intent in Assigned, rather than using it as an early ack, this isn't a concern. `AllNilSubfieldsError` would be encountered prior to checkpointing the record of Assigned config.

File: pkg/registry/core/node/strategy.go [hunk 11]
```diff
@@ -127,14 +127,23 @@ type nodeStatusStrategy struct {
 var StatusStrategy = nodeStatusStrategy{Strategy}
 
 func (nodeStatusStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
-	_ = obj.(*api.Node)
+	node := obj.(*api.Node)
 	// Nodes allow *all* fields, including status, to be set on create.
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DynamicKubeletConfig) {
```
Comment (dashpole) [inline pkg/registry/core/node/strategy.go]: since I am not familiar with the pgk/registry code-base, why is this required?  Same question for addition in PrepareForUpdate.

**Comment by dchen1107** [comment]
I approve the pr but rely on dashpole@ for a throughout review.

/approve

File: pkg/registry/core/node/strategy.go (hunk 11 above)
Comment (mtaufen) [reply pkg/registry/core/node/strategy.go]: We strip alpha fields when the corresponding feature gate is turned off, so that they can't be written unless the feature gate is turned on. See similar code in `func (nodeStrategy) PrepareForCreate/PrepareForUpdate`, similarly see `DropDisabledAlphaFields` in `pkg/api/pod/util.go`.

Comment (dashpole) [reply pkg/registry/core/node/strategy.go]: great, thanks

File: pkg/kubelet/kubeletconfig/status/status.go [hunk 12]
```diff
@@ -19,298 +19,170 @@ package status
...
-func (c *configOkCondition) pokeSyncWorker() {
+func (s *nodeConfigStatus) SetSyncError(err string) {
+	s.transact(func() {
+		s.syncError = fmt.Sprintf(SyncErrorFmt, err)
+	})
+}
+
+func (s *nodeConfigStatus) ClearSyncError() {
+	s.transact(func() {
+		s.syncError = ""
```
[the remark this answers is not in the document]
Comment (dashpole) [reply pkg/kubelet/kubeletconfig/status/status.go]: oh, duh, I get it.

File: pkg/kubelet/kubeletconfig/status/status.go [hunk 13]
```diff
@@ -19,298 +19,170 @@ package status
...
+// NodeConfigStatus represents Node.Status.Config
+type NodeConfigStatus interface {
+	// SetActive sets the active source in the status
+	SetActive(source *apiv1.NodeConfigSource)
+	// SetAssigned sets the assigned source in the status
+	SetAssigned(source *apiv1.NodeConfigSource)
+	// SetLastKnownGood sets the last-known-good source in the status
+	SetLastKnownGood(source *apiv1.NodeConfigSource)
+	// SetError sets the error associated with the status
+	SetError(err string)
```
Comment (dashpole) [inline pkg/kubelet/kubeletconfig/status/status.go]: I would prefer having Error and SyncError functions behave similarly, and I prefer qualifying the error.  I.E.
`SetLoadError`
`ClearLoadError`
`SetSyncError`
`ClearSyncError`
Or just `Set...Error` functions.

The fact that we store the Load error in the status, and override it with the sync error is an implementation detail. (You dont need to call it load error.  It is just what I picked since it comes from loadConfig(assigned).)

Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/status/status.go]: I could see renaming `SetSyncError` to something like `SetTmpError`. I don't think the status manager should assume anything about the context of the base error, which is why I just stuck with `SetError`.

Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/status/status.go]: Updated naming, and simplified to just `SetError` and `SetErrorOverride`.
